pytest/tests/sanity/state_parts_dump_check.py

```diff
@@ -0,0 +1,126 @@
+            )
+            last_height = height
+
+
+def main():
+    node_config_dump, node_config = state_sync_lib.get_state_sync_configs_pair(
+    )
+    config = load_config()
+    near_root, node_dirs = init_cluster(1, 2, 4, config,
+                                        [["epoch_length", EPOCH_LENGTH]], {
+                                            0: node_config,
+                                            1: node_config_dump,
+                                            2: node_config
+                                        })
+
+    boot_node = spin_up_node(config, near_root, node_dirs[0], 0)
+    logger.info('Started boot_node')
+    dump_node = spin_up_node(config,
+                             near_root,
+                             node_dirs[1],
+                             1,
+                             boot_node=boot_node)
+    dump_check = spin_up_node(config,
+                              near_root,
+                              node_dirs[2],
+                              2,
+                              boot_node=boot_node)
+    dump_check.kill()
+    chain_id = boot_node.get_status()['chain_id']
+    dump_folder = node_config_dump["state_sync"]["dump"]["location"]["Filesystem"]["root_dir"]
+    rpc_address, rpc_port = boot_node.rpc_addr()
+    local_rpc_address, local_rpc_port = dump_check.rpc_addr()
+    logger.info(f"{rpc_address}:{rpc_port}")
+    dump_check.run_cmd(
+        cmd=(
+            'state-parts-dump-check',
+            '--chain-id',
+            chain_id,
+            '--root-dir',
+            dump_folder,
+            'loop-check',
+            '--rpc-server-addr',
+            f"http://{rpc_address}:{rpc_port}",
+            '--prometheus-addr',
+            f"{local_rpc_address}:{local_rpc_port}",
+            '--interval',
+            '2'))
+
+    logger.info('Started nodes')
+
+    wait_until(EPOCH_LENGTH - 5, boot_node)
+    # kill dumper node so that it can't dump state.
+    node_height = dump_node.get_latest_block().height
+    logger.info(f'Dump_node@{node_height}')
+    dump_node.kill()
+    logger.info(f'Killed dump_node')
+
+    wait_until(EPOCH_LENGTH + 5, boot_node)
+    # Check the dumped stats.
+    metrics = get_dump_check_metrics(dump_check)
+    assert sum([val for metric, val in metrics.items()
+                if 'state_sync_dump_check_process_is_up' in metric]) == 1
+    assert sum([val for metric, val in metrics.items()
+                if 'state_sync_dump_check_num_parts_valid' in metric]) == 0
+    assert sum([val for metric, val in metrics.items()
+                if 'state_sync_dump_check_num_header_dumped' in metric]) == 0
+
+    wait_until(int(EPOCH_LENGTH + 10), boot_node)
+    # Start state dumper node to keep up with the network and dump state next
+    # epoch.
+    logger.info(f'Starting dump_node')
+    dump_node.start(boot_node=boot_node)
+
+    wait_until(int(EPOCH_LENGTH * 2 + 10), boot_node)
```

Comment:
There's a `utils.poll_epochs()` function that I think would be more precise here. In the past there have been some pytest failures where we make assumptions about where epoch boundaries are, and this is generally not something you're going to get accuracy on by waiting for a particular block height as is done here. Also, do we want to be polling `boot_node` or `dump_node` here? Seems like `dump_node` makes more sense right?

Reply:
I gave `poll_epochs` a go and it it off by 15 blocks. Is too complex for this use case. The function looks at all the validators in the network and checks their epoch and block height and then approximates a timeout. This test only has a validator that generates blocks and we query that validator to see where the chain is. In case of intensive load on the hosts, the dumping node can fall behind and not dump the parts on time but it is not the case here. There is minimum workload on the hosts. Also, the same logic is used in state sync tests and I don't recall any flakiness.

Reply:
Not going to push this point because it's not so important I guess, but I think you're misreading what `poll_epochs()` does. It doesn't look at all validators in the network. It just calls the `validators` rpc method to fetch the current epoch height, and will yield new epoch heights. So since what you're trying to do in this part of the test is wait until a certain epoch is reached, that's exactly what you're looking for.

Reply:
here is a change on top of this PR that seems to work for me. All we want is to get to epoch 2 and then wait a bit right?

```diff
diff --git a/pytest/tests/sanity/state_parts_dump_check.py b/pytest/tests/sanity/state_parts_dump_check.py
index 84aeaf765..ed0f48772 100644
--- a/pytest/tests/sanity/state_parts_dump_check.py
+++ b/pytest/tests/sanity/state_parts_dump_check.py
@@ -13,7 +13,7 @@ import re
 
 sys.path.append(str(pathlib.Path(__file__).resolve().parents[2] / 'lib'))
 
-from utils import wait_for_blocks
+from utils import wait_for_blocks, poll_blocks, poll_epochs
 from cluster import init_cluster, spin_up_node, load_config
 import state_sync_lib
 from configured_logger import logger
@@ -102,7 +102,16 @@ def main():
     logger.info(f'Starting dump_node')
     dump_node.start(boot_node=boot_node)
 
-    wait_for_blocks(node=boot_node, target=EPOCH_LENGTH * 2 + 10)
+    for epoch_height in poll_epochs(boot_node,
+                                    epoch_length=EPOCH_LENGTH):
+        logger.info(f'reached epoch height {epoch_height}')
+        if epoch_height >= 2:
+            break
+    blocks_seen = 0
+    for height_, hash_ in poll_blocks(boot_node):
+        blocks_seen += 1
+        if blocks_seen > 5:
+            break
     # State should have been dumped and reported as dumped.
     metrics = get_dump_check_metrics(dump_check)
     assert sum([val for metric, val in metrics.items(
```

and you could consider doing the same for any other `wait_for_blocks` call whose purpose is also really just to wait for a particular epoch height

Reply:
Don't get me wrong, poll epoch works, it does yield the new epoch but in my case, instead of waiting until block around #100, it stopped at block #115. Both heights are fine, both approaches are fine and spending more time on this is overkill. If it is not just a NIT, I will amend with the suggested changes.